Keyset man (#128)
NLnetLabs/dnst: doc/manual/source/man/dnst-keyset.rst

executing.

For this purpose, keyset has a cron subcommand.
This subcommand handles any house keeping work that needs to be done.
The cron subcommand can either be executed at regular times, for example,
once an hour from the cron(1) utility.

However, keyset also maintains a field in that state file, called
``cron-next``, that specifies when the cron subcommand should be run next.
Running the cron subcommand early is fine, the current time is compared
again the ``cron-next`` field and the subcommand exits early if
``cron-next`` is in the future. 
Running the cron subcommand late may cause signatures to expire.

Create / Init
~~~~~~~~~~~~~

The initialisation of a key set for a zone consists of two steps.
First the create subcommand create a configuration file with mostly default
values and state file without any keys.
The init subcommand finishes the initialisation.

This two step procedure allows configuration parameters to be set between
the create and the init subcommand, for example, the algorithm to use.
It also allows existing public/private key pairs to be imported.

The init subcommand checks if any public/private key pairs have been imported.

> If so, init checks if both a both rolls (KSK and ZSK) are present.

Author: roles

have a SOA serial that is greater or equal to the one that was checked.

Automation of ``cache-expired1`` and ``cache-expired2`` is enabled by the
``expire`` boolean.
When enabled, the cron subcommand simply checks if enough time has passed
to invoke ``cache-expired1`` or ``cache-expired2``.

Finally the ``done`` boolean enabled automation of the ``roll-done`` step.
This automation is very similar to the ``report`` automation.
This only difference is that the Wait actions are automated so propagation
is track but no TTL is reported.

Fine grained control of over automation makes it possible to automate
KSK or algorithm except that they are start manually.
Or let a key roll progress automatically except that the ``cache-expired``
steps are manual to be able insert extra manual steps.

The ``report`` and ``done`` automations require that keyset has network access
to all nameservers of the zone and all nameservers of the parent.

HSM Support (KMIP)
~~~~~~~~~~~~~~~~~~

Importing Keys
~~~~~~~~~~~~~~

There are three basic ways to import exiting keys: public-key,
a public/private key pair from files or a public/private key pair in an HSM.

> A public key can only be import from a file.
> When the key is imported the name of the file is stored in the key set and
> the key will be included in DNSKEY RRset.
> This is useful for certain migration and to manually implement a 
> multi-signer DNSSEC signing setup.
> Note that automation does not work for the case.

Reviewer: ```suggestion
A public key can only be imported from a file.
When the key is imported the name of the file is stored in the keyset state file and
the key will be included in the DNSKEY RRset.
This is useful for certain migrations and to manually implement a 
multi-signer DNSSEC signing setup.
Note that automation does not work for this case.
```

> A public/private key pair can be imported from files.
> It sufficient to give the name of the file that holds the public key if
> the filename ends in ``.key`` and the filename is the private key is the
> same except that it ends in ``.private``. 
> If this is not the case then the private key filename can be specified 
> separately.

Reviewer: ```suggestion
A public/private key pair can be imported from files.
It is sufficient to give the name of the file that holds the public key if
the filename ends in ``.key`` and the filename of the private key is the
same except that it ends in ``.private``. 
If this is not the case then the private key filename must be specified 
separately.
```

> Importing a public/private key on an HSM require specifying the KMIP
> server ID, the ID of the public key, the ID of the private key, the
> DNSSEC algorithm of the key and the flags (typically 256 for a ZSK and
> 257 for a KSK).

Reviewer: ```suggestion
In order to use keys stored in a HSM the ``dnst keyset kmip add-server`` subcommand must first be used to associate the KMIP server connection settings with a user defined server ID.

The first server defined becomes the default. if a default KMIP server has been defined it will be used to generate all future keys, unless the ``dnst keyset kmip disable`` command is issued. If more than one KMIP server is defined, only one can be the default server at any time. Use the ``dnst keyset kmip set-default`` command to change which KMIP server will be used to generate future keys. Note that like all ``dnst keyset` subcommands, the KMIP subcommands set behaviour for a single zone. Additionally there are ``list-servers``, ``get-server``, ``modify-server`` and ``remove-server`` subcommands for inspecting and altering the configured KMIP server settings.

Importing a public/private key stored in an HSM requires specifying the KMIP
server ID, the ID of the public key, the ID of the private key, the
DNSSEC algorithm of the key and the flags (typically 256 for a ZSK and
257 for a KSK).

```

> Normally, keyset assumes ownership of any keys it hold.
> This mean that when a key is deleted from the key set, the keyset subcommand
> will also delete the files that hold the public and private or delete the
> key from the HSM.

Reviewer: ```suggestion
Normally, keyset assumes ownership of any keys it holds.
This means that when a key is deleted from the key set, the keyset subcommand
will also delete the files that hold the public and private keys or delete the
keys from the HSM that was used to create them.
```

> For an import public/private key pair this is considered too dangerous
> because another signer may need the keys.
> For this reason keys are imported in so-called ``decoupled`` state.
> When a decoupled key is deleted, only the reference to the key is deleted
> from the key set, the underlying keys are left untouched.
> There is a ``--coupled`` option to tell keyset to take ownership of the key.

Reviewer: ```suggestion
For an imported public/private key pair this is considered too dangerous
because another signer may need the keys.
For this reason keys are imported in so-called ``decoupled`` state.
When a decoupled key is deleted, only the reference to the key is deleted
from the key set, the underlying keys are left untouched.
There is a ``--coupled`` option to tell keyset to take ownership of the key.
```

key from the HSM.

For an import public/private key pair this is considered too dangerous
because another signer may need the keys.
For this reason keys are imported in so-called ``decoupled`` state.
When a decoupled key is deleted, only the reference to the key is deleted
from the key set, the underlying keys are left untouched.
There is a ``--coupled`` option to tell keyset to take ownership of the key.

Migration
~~~~~~~~~

The keyset subcommand has no direct support for migration.
Migration has to be done manually using the import commands.
The semeantics of the import commands is decribed in the previous section.
This section focusses on how the import command can be used to perform a
migration.

There are three migration strategies: 1) importing the existing signer's
(private) signing keys, 2) a full multi-signer migration and 3)
a partial multi-signer migration.

Importing the existing signer's signing keys
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Importing the existing signer's public/private keys pairs is the easiest
migration mechanism.
The basic process is the following:

* Disable (automatic) key rolls on the existing signer

> * Disable automatic key rolls before the executing the create command.

Reviewer: ```suggestion
* Disable automatic key rolls before executing the create command.
```

* Import the KSK and ZSK (or CSK) as files or using KMIP between the
  create and init commands.

* Check with tools such as ldns-verify-zone that the new zone is secure with
  the existing DS record at the parent.

> * Switch the nameservers to the new signer.

Reviewer: Do you mean ensure that hidden primary nameserver zone content is delivered to the new signer for signing, whether that be via file copying scripts or shared filesystem or reconfiguring the hidden primary nameserver(s) for the zone to send NOTIFY to, and respond to XFR from, the new (name)server that will be doing the signing?

* Perform key rolls for the KSK and ZSK (or the CSK).

> * (If needed) enable automatic key rolls.

Reviewer: Should that be wanted rather than needed?

Author: Yes

> * Remove the zone from the old signer.

Reviewer: Why? Do you mean disable propagation of the old signer output zone?

> Note that after the key roll, to signer have access to the signing keys.

Reviewer: There's something wrong with this sentence.

In case of KMIP keys, the old signer can also delete the keys from the HSM.
For this reason it is best to perform key rolls of all keys before removing
the zone from the old signer.

This document describes key management. Care should be taken that other

> parameters, such as the use of NSEC or NSEC3 and NSEC3 parameters are

Reviewer: NSEC3 twice.

the same (to avoid confusion) and that the SOA serial policy is the same
(to avoid problems with zone transfers).

Full multi-signer migration
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The basic idea is the execute the following steps:

Reviewer: ```suggestion
The basic idea is to execute the following steps:
```

* Disable (automatic) key rolls on the existing signer.

> * If the parent supports automatic updating the DS record using CDS/CDNSKEY
>   (RFC 8078) then disable the generation of CDS/CDNSKEY records on the
>   existing or disable CDS/CDNSKEY processing for this zone at the parent.

Reviewer: ```suggestion
* If the parent supports automatic updating of the DS record using CDS/CDNSKEY
  (RFC 8078) then disable the generation of CDS/CDNSKEY records on the
  existing signer or disable CDS/CDNSKEY processing for this zone at the parent.
```

* Issue the create command.

* Disable automatic key rolls.

> * (Disable CDS/CDNSKEY generation. Keyset does not have that at the moment)

Reviewer: Keyset doesn't have disable, or doesn't generate them at all?

> * Import the public key of the existing signer's ZSK (or CSK).

Reviewer: Using a keyset command?

* Issue the init command.

* Make sure in the next step to only add a DS record at the parent, not

>   the delete the existing one.

Reviewer: ```suggestion
  delete the existing one.
```

* Complete the initial algorithm roll.

* Verify using tools such as ldns-veridy-zone that the zone is correctly
  signed.

* Import the public key of the new ZSK (or CSK) in the existing signer.

> * Verify that all nameservers that serve the zone have new ZSK in the

Reviewer: ```suggestion
* Verify that all nameservers that serve the zone have the new ZSK in the
```

  DNSKEY RRset of the existing signer.

* Transition the nameservers from the existing signer to the new signer.

* Let caches expire for the DNSKEY RRset of the old signer and the
  zone RRSIGs of the old signer.

* Remove the DS record for the old signer from the parent.

* Remove the imported public key.

> * (If needed) enable automatic key rolls and generation of CDS/CDNSKEY

Reviewer: Needed, or wanted?

Author: changed to wanted

  records.

Partial multi-signer migration
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

A partial multi-signer migration is the right approach when the existing
signer cannot import the new signers ZSK.
A requirement is that the new signer can transfer the signed zone from the

> existing signer and that the new signer supports to so-called "pass-through"

Reviewer: ```suggestion
existing signer and that the new signer supports so-called "pass-through"
```

mode.
In pass-through mode a signer leaves signatures for zone records unchanged

> be does replace the DNSKEY, CDS and CDNSKEY RRset with the ones from

Reviewer: ```suggestion
but does replace the DNSKEY, CDS and CDNSKEY RRset with the ones from
```

this subcommand.

> The basic idea is the execute the following steps:

Reviewer: ```suggestion
The basic idea is to execute the following steps:
```

* Disable (automatic) key rolls on the existing signer.

> * If the parent supports automatic updating the DS record using CDS/CDNSKEY
>   (RFC 8078) then disable the generation of CDS/CDNSKEY records on the
>   existing or disable CDS/CDNSKEY processing for this zone at the parent.

Reviewer: ```suggestion
* If the parent supports automatic updating of the DS record using CDS/CDNSKEY
  (RFC 8078) then disable the generation of CDS/CDNSKEY records in the
  existing signer or disable CDS/CDNSKEY processing for this zone at the parent.
```

* Issue the create command.

* Disable automatic key rolls.

> * (Disable CDS/CDNSKEY generation. Keyset does not have that at the moment)

Reviewer: Same comment as above.

* Import the public key of the existing signer's ZSK (or CSK).

* Issue the init command.

> * Switch the signer to pass-through mode. The signer has to transfer the

Reviewer: ```suggestion
* Switch the new signer to pass-through mode. The signer has to transfer the
```

  signed zone from the existing signer.

* Make sure in the next step to only add a DS record at the parent, not
  the delete the existing one.

* Complete the initial algorithm roll.

* Verify using tools such as ldns-veridy-zone that the zone is correctly
  signed.

* Transition the nameservers from the existing signer to the new signer.

* Let caches expire for the DNSKEY RRset of the old signer.

* Remove the DS record for the old signer from the parent.

> * switch off pass-through mode.

Reviewer: ```suggestion
* Switch off pass-through mode.
```
